File: content/snippets/python/interview_lcci_17_15_longest_word.py
from typing import List


class Solution:
    def longestWord(self, words: List[str]) -> str:
        word_set = set(words)

        def can_build(w: str) -> bool:
            if not w:
                return True
            for i in range(1, len(w) + 1):
                if w[:i] in word_set and can_build(w[i:]):
                    return True
            return False

        words.sort(key=lambda x: (-len(x), x))
        for w in words:
            word_set.discard(w)
            if can_build(w):
                return w
            word_set.add(w)
        return ""

File: content/snippets/scala2/interview_lcci_17_15_longest_word.scala
object Solution {
  def longestWord(words: Array[String]): String = {
    val wordSet = scala.collection.mutable.Set(words: _*)

    def canBuild(w: String): Boolean = {
      if (w.isEmpty) return true
      (1 to w.length).exists(i => wordSet.contains(w.take(i)) && canBuild(w.drop(i)))
    }

    val sorted = words.sortWith((a, b) => if (a.length != b.length) a.length > b.length else a < b)
    for (w <- sorted) {
      wordSet.remove(w)
      if (canBuild(w)) return w
      wordSet.add(w)
    }
    ""
  }
}
